### PR TITLE
feat(ids-admin): Delegation-Delegation-Type

### PR DESCRIPTION
**Pull request by GunnlaugurG**
## What

Add `delegation_delegation_type` to db and support General Mandate delegations (Allsherjarumboð) 

## Why

So the new delegations-admin module can create GeneralMandate delegation

## Screenshots / Gifs

Attach Screenshots / Gifs to help reviewers understand the scope of the pull request

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Formatting passes locally with my changes
- [ ] I have rebased against main before asking for a review

## Summary by CodeRabbit

- **New Features**
	- Introduced support for "General Mandate" delegation type across various components and services.
	- Enhanced delegation handling, allowing for improved retrieval and management of delegation data.
	- Added new methods to fetch valid and available general mandates based on national IDs.

- **Bug Fixes**
	- Improved test coverage for delegation functionalities, ensuring accurate behavior of API endpoints related to general mandates.

### CODE REVIEW

**Review by jonarnarbriem — APPROVED** [review]

File: libs/auth-api-lib/src/lib/delegations/delegation-scope.service.ts [hunk 1]
```diff
@@ -161,6 +167,47 @@ export class DelegationScopeService {
     })
   }
 
+  private findValidGeneralMandateTo(
+    toNationalId: string,
+    fromNationalId: string,
+  ): Promise<Delegation[]> {
```
Comment (saevarma) [inline libs/auth-api-lib/src/lib/delegations/delegation-scope.service.ts:173]: We should let this be focused on finding scopes

```suggestion
  private findValidGeneralMandateScopesTo(
    toNationalId: string,
    fromNationalId: string,
  ): Promise<String[]> {
```

File: libs/auth-api-lib/src/lib/delegations/delegation-scope.service.ts [hunk 2]
```diff
@@ -161,6 +167,47 @@ export class DelegationScopeService {
...
+                  where: {
+                    enabled: true,
+                  },
+                },
+              ],
+            },
+          ],
+        },
+      ],
+    })
```
Comment (saevarma) [inline libs/auth-api-lib/src/lib/delegations/delegation-scope.service.ts:208]: I think its better to split this into:

1. Query `DelegationModel` to check if a valid GeneralMandate delegation is found. If not found return empty array
2. If valid then find all api scopes that has GeneralMandate relation.

File: libs/auth-api-lib/src/lib/delegations/delegations-incoming.service.ts [hunk 3]
```diff
@@ -182,6 +188,17 @@ export class DelegationsIncomingService {
       )
     }
 
+    // This has a provider of Custom but needs to be handled separately
```
Comment (saevarma) [inline libs/auth-api-lib/src/lib/delegations/delegations-incoming.service.ts:191]: Custom is also provider of GeneralMandate, we should explicitly check the custom delegation type in the if above.

File: libs/auth-api-lib/src/lib/delegations/models/delegation-delegation-type.model.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  DataType,
+  CreatedAt,
+  UpdatedAt,
+  BelongsTo,
+} from 'sequelize-typescript'
+
+import { Delegation } from './delegation.model'
+import { DelegationTypeModel } from './delegation-type.model'
+
+@Table({
```
Comment (saevarma) [inline libs/auth-api-lib/src/lib/delegations/models/delegation-delegation-type.model.ts:20]: You can and should add the delegation delegation_type unique constraint in the `@Table` decorator, so tests run with the same constraints as prod.

File: libs/portals/shared-modules/delegations/src/lib/messages.ts
Comment (saevarma) [inline libs/portals/shared-modules/delegations/src/lib/messages.ts:1]: Remember to extract-strings and add the translation in Contentful.

File: libs/auth-api-lib/src/lib/delegations/delegations-incoming-custom.service.ts [hunk 5]
```diff
@@ -84,6 +83,33 @@ export class DelegationsIncomingCustomService {
...
+          validity: DelegationValidity.INCLUDE_FUTURE,
+        },
+        useMaster,
+      )
+
+    const delegationDTOs = delegations.map((d) =>
+      d.toDTO(AuthDelegationType.GeneralMandate),
+    )
+
+    return delegationDTOs.map((d) => {
```
Comment (saevarma) [inline libs/auth-api-lib/src/lib/delegations/delegations-incoming-custom.service.ts:103]: Here we are looping 2x through the same array, we should join these maps to safe the extra iteration.

File: libs/auth-api-lib/src/lib/delegations/delegations-incoming-custom.service.ts [hunk 6]
```diff
@@ -151,6 +178,117 @@ export class DelegationsIncomingCustomService {
...
+      nationalId,
+      validity,
+    }: FindAllValidIncomingOptions & {
+      validity: DelegationValidity
+    },
+    useMaster = false,
+  ): Promise<{ delegations: Delegation[]; fromNameInfo: FromNameInfo[] }> {
+    const whereOptions = getScopeValidityWhereClause(validity)
+
+    const delegations = await this.delegationModel.findAll({
```
Comment (saevarma) [inline libs/auth-api-lib/src/lib/delegations/delegations-incoming-custom.service.ts:230]: We don't need to the join with ApiScope

File: libs/auth-api-lib/src/lib/delegations/delegations-incoming-custom.service.ts [hunk 7]
```diff
@@ -151,6 +178,117 @@ export class DelegationsIncomingCustomService {
...
+  private async findAllIncomingGeneralMandates(
+    {
+      nationalId,
+      validity,
+    }: FindAllValidIncomingOptions & {
+      validity: DelegationValidity
+    },
+    useMaster = false,
+  ): Promise<{ delegations: Delegation[]; fromNameInfo: FromNameInfo[] }> {
+    const whereOptions = getScopeValidityWhereClause(validity)
```
Comment (saevarma) [inline libs/auth-api-lib/src/lib/delegations/delegations-incoming-custom.service.ts:228]: We are not using the `validFrom` fields for GeneralMandates so this becomes a bit weird. As `findAllValidGeneralMandates` uses `INCLUDE_FUTURE` which sounds strange as it should return only valid `NOW`. But you are doing that to work   around the `validFrom` check introduced by this helper.

So I would suggest to not use this helper here, and just arrange the `whereOptions` here to handle `ALL` and `NOW`

**Review by saevarma — APPROVED** [review]
🚀

**Pull request merged**